Sources/NIOCore/GlobalSingletons.swift

```diff
@@ -0,0 +1,180 @@
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#else
+#error("Unsupported C library")
+#endif
+
+/// SwiftNIO provided singleton resources for programs & libraries that don't need full control over all operating
+/// system resources. This type holds sizing (how many loops/threads) suggestions.
+///
+/// Users who need very tight control about the exact threads and resources created may decide to set
+/// `NIOGlobalSingletons.globalSingletonsEnabledSuggestion = false`. All singleton-creating facilities should check
+/// this setting and if `false` restrain from creating any global singleton resources. Please note that disabling the
+/// global singletons will lead to a crash if _any_ code attempts to use any of the singletons.
+public enum NIOGlobalSingletons {
+}
+
+extension NIOGlobalSingletons {
+    /// A suggestion of how many ``EventLoop``s the global singleton ``EventLoopGroup``s are supposed to consist of.
+    ///
+    /// The thread count is ``System/coreCount`` unless the environment variable `NIO_SINGLETON_GROUP_LOOP_COUNT`
+    /// is set or this value was set manually by the user.
+    ///
+    /// - note: This value must be set _before_ any singletons are used and must only be set once.
+    public static var suggestedGlobalSingletonGroupLoopCount: Int {
+        set {
+            Self.userSetSingletonThreadCount(rawStorage: globalRawSuggestedLoopCount, userValue: newValue)
+        }
+
+        get {
+            return Self.getTrustworthyThreadCount(rawStorage: globalRawSuggestedLoopCount,
+                                                  environmentVariable: "NIO_SINGLETON_GROUP_LOOP_COUNT")
+        }
+    }
+
+    /// A suggestion of how many threads the global singleton thread pools that can be used for synchronous, blocking
+    /// functions (such as ``NIOThreadPool``) are supposed to consist of
+    ///
+    /// The thread count is ``System/coreCount`` unless the environment variable
+    /// `NIO_SINGLETON_BLOCKING_POOL_THREAD_COUNT` is set or this value was set manually by the user.
+    ///
+    /// - note: This value must be set _before_ any singletons are used and must only be set once.
+    public static var suggestedBlockingPoolThreadCount: Int {
+        set {
+            Self.userSetSingletonThreadCount(rawStorage: globalRawSuggestedBlockingThreadCount, userValue: newValue)
+        }
+
+        get {
+            return Self.getTrustworthyThreadCount(rawStorage: globalRawSuggestedBlockingThreadCount,
+                                                  environmentVariable: "NIO_SINGLETON_BLOCKING_POOL_THREAD_COUNT")
+        }
+    }
+
+    /// A suggestion for whether the global singletons should be enabled. This is `true` unless changed by the user.
+    ///
+    /// This value cannot be changed using an environment variable.
+    ///
+    /// - note: This value must be set _before_ any singletons are used and must only be set once.
+    public static var globalSingletonsEnabledSuggestion: Bool {
```

Comment:
Is this effectively a kill-switch for singletons? I.e. set this and blow up if singletons are used?

Reply:
@glbrntt It's a suggestion because we can't _actually_ control that no code's creating singletons... We can make NIO code behave but we don't actually know :)

Reply:
kill switch: Yes, I think this is important to test stuff that thinks/hopes/promises to correctly wire through the `group`. Without the kill switch it's super duper hard to test (need to check the thread count etc)